frame/vamm/src/helpers/twap.rs

```diff
@@ -5,29 +5,49 @@ use sp_runtime::traits::Saturating;
 
 impl<T: Config> Pallet<T> {
 	/// Performs runtime storage changes, effectively updating the asset twap.
+	/// This `update_twap` variation can't accept any error to occur, expecting
+	/// all properties described in
+	/// [`update_twap`](struct.Pallet.html#method.update_twap) to hold.
 	///
 	/// # Errors
 	///
 	/// * All errors returned by
-	/// [`sanity_check_before_update_twap`](Self::sanity_check_before_update_twap).
+	/// [`sanity_check_before_update_twap`](
+	/// struct.Pallet.html#method.sanity_check_before_update_twap).
 	#[transactional]
 	pub fn do_update_twap(
 		vamm_id: T::VammId,
 		vamm_state: &mut VammStateOf<T>,
-		base_twap: T::Decimal,
+		base_twap: Option<T::Decimal>,
 		now: &Option<T::Moment>,
 	) -> Result<T::Decimal, DispatchError> {
-		// Sanity checks must pass before updating runtime storage.
-		Self::sanity_check_before_update_twap(vamm_state, base_twap, now)?;
-
-		let now = Self::now(now);
-		vamm_state.base_asset_twap = base_twap;
-		vamm_state.twap_timestamp = now;
-
-		// Update runtime storage.
-		VammMap::<T>::insert(&vamm_id, vamm_state);
+		Self::internal_update_twap(vamm_id, vamm_state, base_twap, now, false)
+	}
 
-		Ok(base_twap)
+	/// *Tries to* perform runtime storage changes, effectively updating the
+	/// asset twap.  Diferently than [`do_update_twap`](Self::do_update_twap) though, this variation
+	/// accepts one specific error to occur, therefore not honoring all
+	/// properties described in
+	/// [`update_twap`](../pallet/struct.Pallet.html#method.update_twap). The
+	/// property that can be violated using this variation is that the current
+	/// twap timestamp can in fact be more recent than the current time.
+	///
+	/// # Errors
+	///
+	/// * All errors returned by
+	/// [`sanity_check_before_update_twap`](
+	/// struct.Pallet.html#method.sanity_check_before_update_twap),
+	/// except
+	/// [`Error::<T>::AssetTwapTimestampIsMoreRecent`](
+	/// ../pallet/enum.Error.html#variant.AssetTwapTimestampIsMoreRecent).
+	#[transactional]
```

Comment:
What's the Substrate version that we're using? There's already one which includes a transactional layer by default.